Service/EnvironmentService.php

```diff
@@ -32,24 +41,106 @@ class EnvironmentService
     /** @var AuthorizationCheckerInterface $authorizationChecker*/
     private $authorizationChecker;
 
+    /**@var Container $container*/
+    private $container;
+
+    /** @var Logger */
+    private $logger;
+
+    /**@var Client */
+    private $client;
+
+    /** @var ContentTypeService
+    private $contentTypeService;
+     * */
+
     private $singleTypeIndex;
 
-    public function __construct(Registry $doctrine, Session $session, UserService $userService, AuthorizationCheckerInterface $authorizationChecker, $singleTypeIndex)
-    {
+    public function __construct(
+        Registry $doctrine,
+        Session $session,
+        UserService $userService,
+        AuthorizationCheckerInterface $authorizationChecker,
+        Container $container,
+        Logger $logger,
+        Client $client,
+        $singleTypeIndex
+    ) {
         $this->doctrine = $doctrine;
         $this->session = $session;
         $this->userService = $userService;
         $this->authorizationChecker = $authorizationChecker;
+        $this->container = $container;
+        $this->logger = $logger;
+        $this->client = $client;
         $this->singleTypeIndex = $singleTypeIndex;
     }
 
+    public function createEnvironment(string $name, $snapshot = false): Environment
+    {
+        if (!$this->validateEnvironmentName($name)) {
+            throw new \Exception('An environment name must respects the following regex /^[a-z][a-z0-9\-_]*$/');
+        }
+
+        $environment = new Environment();
+        $environment->setName($name);
+        $environment->setAlias($this->container->getParameter('ems_core.instance_id') . $environment->getName());
+        $environment->setManaged(true);
+        $environment->setSnapshot($snapshot);
+
+        try {
+            $em = $this->doctrine->getManager();
+            $em->persist($environment);
+            $em->flush();
+        } catch (\Exception $e) {
+            dump($e->getMessage());
+        }
+
+        $indexName = $environment->getAlias() . AppController::getFormatedTimestamp();
+        $this->client->indices()->create([
+            'index' => $indexName,
+            'body' => $this->getIndexAnalysisConfiguration(),
+        ]);
+
+        $contentTypeService = $this->container->get('ems.service.contenttype');
```

Comment:
Injection via Dependency Injection? container->get is deprecated!

Reply:
Yes but no : if we use D.I. there is a Circular reference error (ems.service.contenttype has ems.service.environement as dependency)

Reply:
Interesting (this is not good! :D )